src/services/pcn-bridge/src/Bridge_dp.c

```diff
@@ -0,0 +1,357 @@
+ */
+struct port_vlan_key {
+  u32 port;
+  u16 vlan;
+} __attribute__((packed));
+
+struct port_vlan_value {
+  u16 vlan;
+  u16 stp_state;
+} __attribute__((packed));
+
+BPF_TABLE("hash", struct port_vlan_key, struct port_vlan_value, port_vlan,
+          1024);
+/*
+ * port_vlan usage example:
+ * <port> <vlan> -> vlanid
+ *
+ * port 1 access vlan=10
+ * 1 VLAN_WILDCARD -> 10
+ *
+ * port 3 trunk vlan=30,40
+ * 3 30 -> 30
+ * 3 40 -> 40
+ *
+ * port 4 trunk vlan=all
+ * 4 VLAN_WILDCARD -> VLAN_WILDCARD
+ */
+
+/*
+ * Fowarding Table: VLAN + MAC Address to port association.  This table is
+ * filled
+ * in the learning phase and then is used in the forwarding phase to decide
+ * where to send a packet
+ */
+struct fwd_key {
+  u16 vlan;
+  __be64 mac;
+} __attribute__((packed));
+
+struct fwd_entry {
+  u32 timestamp;
+  u32 port;
+  enum entry_type type;
+} __attribute__((packed));
+
+BPF_TABLE("hash", struct fwd_key, struct fwd_entry, fwdtable, 1024);
+BPF_TABLE("array", int, uint32_t, timestamp, 1);
+
+static __always_inline u32 time_get_sec() {
+  int key = 0;
+  u32 *ts = timestamp.lookup(&key);
+  if (ts)
+    return *ts;
+
+  return 0;
+}
+
+/*
+ * Port Table: Saves the state of the ports
+ * TODO: what is the best option in this case, hash vs array?
+ */
+struct port {
+  u16 mode;
+  u16 native_vlan;
+  bool native_vlan_enabled;
+};
+
+BPF_TABLE("hash", u32, struct port, ports, 256);
+
+struct eth_hdr {
+  __be64 dst : 48;
+  __be64 src : 48;
+  __be16 proto;
+} __attribute__((packed));
+
+static __always_inline int handle_rx(struct CTXTYPE *ctx,
+                                     struct pkt_metadata *md) {
+  void *data = (void *)(long)ctx->data;
+  void *data_end = (void *)(long)ctx->data_end;
+  struct eth_hdr *eth = data;
+
+  if (data + sizeof(*eth) > data_end)
+    goto DROP;
+
+  u32 in_ifc = md->in_port;
+
+  pcn_log(ctx, LOG_DEBUG, "in_ifc=%d proto=0x%x", in_ifc,
+          bpf_ntohs(eth->proto));
+  pcn_log(ctx, LOG_DEBUG, "mac src:%M dst:%M", eth->src, eth->dst);
+
+  u16 vlanid = 0;
+
+  bool tagged = pcn_is_vlan_present(ctx);
+
+  if (tagged) {
+    int result;
+    if ((result = pcn_get_vlan_id(ctx)) < 0)
+      goto DROP;
+    vlanid = result;
+  }
+
+  struct port *in_port = ports.lookup(&in_ifc);
+  if (!in_port) {
+    pcn_log(ctx, LOG_ERR, "Port %u not found", in_ifc);
+    goto DROP;
+  }
+
+  struct port_vlan_key vlan_key;
+  struct port_vlan_value *vlan_entry;
+
+  vlan_key.port = in_ifc;
+
+  if (in_port->mode == PORT_MODE_ACCESS) {
+    if (tagged)
+      goto DROP;
+
+    vlan_key.vlan = VLAN_WILDCARD;
+    vlan_entry = port_vlan.lookup(&vlan_key);
+    if (!vlan_entry) {
+      pcn_log(ctx, LOG_ERR, "no VLAN configured in this access port");
+      goto DROP;
+    }
+
+    vlanid = vlan_entry->vlan;
+
+    pcn_log(ctx, LOG_TRACE, "port mode ACCESS vlan=%d, state=%d", vlanid,
+            vlan_entry->stp_state);
+  } else {
+    // TRUNK
+    if (!tagged) {
+      vlanid = in_port->native_vlan;
```

Comment:
What happens if `native-vlan-enabled` is `false`?

Reply:
I added a check in order to manage this case.

Reply:
if native-vlan-enabled== false, you entered in previous "if" and dropped the packet.